core/src/cost_model.rs

```diff
@@ -0,0 +1,466 @@
+    pub costliest_account_cost: u32,
+}
+
+#[derive(Debug)]
+pub struct CostModel {
+    cost_metrics: HashMap<Pubkey, u32>,
+    cost_tracker: CostTracker,
+    current_bank_slot: Slot,
+}
+
+macro_rules! costmetrics {
+    ($( $key: expr => $val: expr ),*) => {{
+        let mut hashmap: HashMap< Pubkey, u32 > = HashMap::new();
+        $( hashmap.insert( $key, $val); )*
+        hashmap
+    }}
+}
+
+impl Default for CostModel {
+    fn default() -> Self {
+        CostModel::new()
+    }
+}
+
+impl CostModel {
+    pub fn new() -> Self {
+        Self::new_with_config(CHAIN_MAX_COST, BLOCK_MAX_COST)
+    }
+
+    // returns total block cost if succeeded in adding;
+    pub fn try_to_add_transaction(&mut self, transaction: &Transaction) -> Option<u32> {
+        let writable_accounts = &Self::find_writable_keys(transaction.message())[..];
+        let transaction_cost = self.find_transaction_cost(&transaction);
+
+        if self
+            .cost_tracker
+            .would_exceed_limit(writable_accounts, &transaction_cost)
+        {
+            debug!("can not fit transaction {:?}", transaction);
+            None
+        } else {
+            debug!("transaction {:?} added to block", transaction);
+            self.cost_tracker
+                .add_transaction(writable_accounts, &transaction_cost);
+            Some(*self.cost_tracker.package_cost())
+        }
+    }
+
+    pub fn get_stats(&self) -> CostModelStats {
+        // A temp method to collect bank cost stats
+        let mut stats = CostModelStats {
+            total_cost: *self.cost_tracker.package_cost(),
+            number_of_accounts: self.cost_tracker.account_costs().len(),
+            costliest_account: Pubkey::default(),
+            costliest_account_cost: 0,
+        };
+
+        for (key, cost) in self.cost_tracker.account_costs().iter() {
+            if cost > &stats.costliest_account_cost {
+                stats.costliest_account = *key;
+                stats.costliest_account_cost = *cost;
+            }
+        }
+
+        stats
+    }
+
+    pub fn total_cost(&self) -> &u32 {
+        self.cost_tracker.package_cost()
+    }
+
+    pub fn reset_if_new_bank(&mut self, slot: Slot) {
+        if slot != self.current_bank_slot {
+            self.cost_tracker.reset();
+            self.current_bank_slot = slot;
+        }
+    }
+
+    fn new_with_config(chain_max: u32, block_max: u32) -> Self {
+        debug!(
+            "new cost model with chain_max {}, block_max {}",
+            chain_max, block_max
+        );
+
+        // NOTE: message.rs has following lazy_static program ids.
+        let parse = |s| Pubkey::from_str(s).unwrap();
+        Self {
+            cost_metrics: costmetrics![
+                parse("Config1111111111111111111111111111111111111") => COST_UNIT,
+                parse("Feature111111111111111111111111111111111111") => COST_UNIT * 2,
+                parse("NativeLoader1111111111111111111111111111111") => COST_UNIT * 2,
```

Comment:
Can we not use the hardcoded ones?

There should be an `id(()` for each of these.

Reply:
pushed the change ... thanks for pointing it out